source/extensions/filters/http/ext_authz/ext_authz.cc

```diff
@@ -46,25 +46,28 @@ void Filter::initiateCall(const Http::HeaderMap& headers) {
                                                                 check_request_);
 
   state_ = State::Calling;
+  // Don't let the filter chain continue as we are going to invoke check call.
+  filter_return_ = FilterReturn::StopDecoding;
   initiating_call_ = true;
   client_->check(*this, check_request_, callbacks_->activeSpan());
   initiating_call_ = false;
 }
 
 Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool) {
   initiateCall(headers);
-  return state_ == State::Calling ? Http::FilterHeadersStatus::StopIteration
-                                  : Http::FilterHeadersStatus::Continue;
+  return filter_return_ == FilterReturn::StopDecoding ? Http::FilterHeadersStatus::StopIteration
```

Comment:
Thanks for the fix. IMHO, we have a lot of complexity here, since we now have 3 state variables, and I can't see the clear relationship between them. The variables are `filter_return_`, `initiating_call_` and `state_`. Can you write a comment explaining how they relate? Is there no way to simplify this?